// @payloadcms/ui/dist/elements/DocumentDrawer/DrawerHeader/index.js
'use client';

import { c as _c } from "react/compiler-runtime";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Gutter } from '../../../elements/Gutter/index.js';
import { useModal } from '../../../elements/Modal/index.js';
import { RenderTitle } from '../../../elements/RenderTitle/index.js';
import { XIcon } from '../../../icons/X/index.js';
import { useDocumentInfo } from '../../../providers/DocumentInfo/index.js';
import { useDocumentTitle } from '../../../providers/DocumentTitle/index.js';
import { useTranslation } from '../../../providers/Translation/index.js';
import { IDLabel } from '../../IDLabel/index.js';
import { documentDrawerBaseClass } from '../index.js';
import './index.scss';
export const DocumentDrawerHeader = t0 => {
  const $ = _c(9);
  const {
    AfterHeader,
    drawerSlug,
    showDocumentID: t1
  } = t0;
  const showDocumentID = t1 === undefined ? true : t1;
  const {
    closeModal
  } = useModal();
  const {
    t
  } = useTranslation();
  let t2;
  if ($[0] !== AfterHeader || $[1] !== closeModal || $[2] !== drawerSlug || $[3] !== showDocumentID || $[4] !== t) {
    let t3;
    if ($[6] !== closeModal || $[7] !== drawerSlug) {
      t3 = () => closeModal(drawerSlug);
      $[6] = closeModal;
      $[7] = drawerSlug;
      $[8] = t3;
    } else {
      t3 = $[8];
    }
    t2 = _jsxs(Gutter, {
      className: `${documentDrawerBaseClass}__header`,
      children: [_jsxs("div", {
        className: `${documentDrawerBaseClass}__header-content`,
        children: [_jsx("h2", {
          className: `${documentDrawerBaseClass}__header-text`,
          children: _jsx(RenderTitle, {
            element: "span"
          })
        }), _jsx("button", {
          "aria-label": t("general:close"),
          className: `${documentDrawerBaseClass}__header-close`,
          onClick: t3,
          type: "button",
          children: _jsx(XIcon, {})
        })]
      }), showDocumentID && _jsx(DocumentID, {}), AfterHeader ? _jsx("div", {
        className: `${documentDrawerBaseClass}__after-header`,
        children: AfterHeader
      }) : null]
    });
    $[0] = AfterHeader;
    $[1] = closeModal;
    $[2] = drawerSlug;
    $[3] = showDocumentID;
    $[4] = t;
    $[5] = t2;
  } else {
    t2 = $[5];
  }
  return t2;
};
const DocumentID = () => {
  const $ = _c(3);
  const {
    id
  } = useDocumentInfo();
  const {
    title
  } = useDocumentTitle();
  let t0;
  if ($[0] !== id || $[1] !== title) {
    t0 = id && id !== title ? _jsx(IDLabel, {
      id: id.toString()
    }) : null;
    $[0] = id;
    $[1] = title;
    $[2] = t0;
  } else {
    t0 = $[2];
  }
  return t0;
};
